Vectorize one loop

kronbichler opened the pull request.

bangerth commented on include/deal.II/lac/vector.templates.h:
```diff
@@ -795,11 +795,9 @@ namespace internal
-        *p = *q;
+      DEAL_II_OPENMP_SIMD_PRAGMA
+      for (unsigned int i=begin; i<end; ++i)
+        dst.begin()[i] = src.begin()[i];
```
Do you think that these are equally efficient?

kronbichler commented:
Note that the code I changed is only used for conversion operations, the plain copy operations goes through memcpy (which uses whatever optimized calls are available; on my system through movsb). I checked the assembly output for double->float and float->double and the code is nicely vectorized and unrolled by a factor of 8 with pretty optimal assembler code. I also checked double->complex<double> conversion and the assembler code looks still optimal (though not vectorized of course).
One thing I observed was that debug code is faster if I pull out the begin() statements from the loop because the compiler will not optimize anything), which I changed in a second commit.

In making these tests, the second commit also adds instantiations for copy real -> complex which seem useful (but not the other way around which does not make sense).
